File: hydra-haskell/src/main/haskell/Hydra/Dsl/Meta/Lib/Pairs.hs
-- | Phantom-typed term DSL for the hydra.lib.pairs library

module Hydra.Dsl.Meta.Lib.Pairs where

import Hydra.Phantoms
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries


bimap :: TTerm (a -> c) -> TTerm (b -> d) -> TTerm (a, b) -> TTerm (c, d)
bimap = primitive3 _pairs_bimap

first :: TTerm (a, b) -> TTerm a
first = primitive1 _pairs_first

second :: TTerm (a, b) -> TTerm b
second = primitive1 _pairs_second

File: packages/hydra-java/src/main/java/hydra/dsl/meta/lib/Pairs.java
package hydra.dsl.meta.lib;
import hydra.dsl.meta.Phantoms;
import hydra.typed.TypedTerm;

/** Phantom-typed term DSL for the {@code hydra.lib.pairs} library. */
public final class Pairs {
    private Pairs() {}

    private static <A> TypedTerm<A> prim(String name) {
        return Phantoms.var("hydra.lib.pairs." + name);
    }

    public static <A> TypedTerm<A> bimap(TypedTerm<?> f, TypedTerm<?> g, TypedTerm<?> p) { return Phantoms.apply(prim("bimap"), f, g, p); }
    public static <A> TypedTerm<A> first(TypedTerm<?> pair) { return Phantoms.apply(prim("first"), pair); }
    public static <A> TypedTerm<A> second(TypedTerm<?> pair) { return Phantoms.apply(prim("second"), pair); }
}
